Userland/Applications/PixelPaint/Tools/TextTool.cpp

```diff
@@ -0,0 +1,295 @@
+    m_text_editor->set_font(m_selected_font);
+    m_cursor_blink_timer = Core::Timer::construct();
+    m_cursor_blink_timer->on_timeout = [&]() {
+        m_cursor_blink_state = !m_cursor_blink_state;
+    };
+    m_cursor_blink_timer->set_interval(500);
+}
+
+void TextTool::on_tool_deactivation()
+{
+    reset_tool();
+}
+void TextTool::on_mousemove(Layer*, MouseEvent& event)
+{
+    if (m_text_input_is_active) {
+        auto mouse_position = editor_stroke_position(event.layer_event().position(), 1);
+        m_mouse_is_over_text = m_ants_rect.contains(mouse_position);
+        m_editor->update_tool_cursor();
+    }
+
+    if (m_is_dragging) {
+        auto new_position = event.layer_event().position();
+        m_add_text_position = m_add_text_position + (new_position - m_drag_start_point);
+        m_drag_start_point = new_position;
+    }
+}
+void TextTool::on_mouseup(Layer*, MouseEvent&)
+{
+    m_is_dragging = false;
+}
+void TextTool::on_mousedown(Layer*, MouseEvent& event)
+{
+    auto start_text_region = [&] {
+        m_text_color = m_editor->color_for(event.layer_event());
+        m_text_input_is_active = true;
+        m_text_editor->set_text(""sv);
+        m_add_text_position = event.layer_event().position();
+        m_editor->image().selection().begin_interactive_selection();
+        m_cursor_blink_timer->start();
+        m_editor->update();
+    };
+
+    if (!m_text_input_is_active) {
+        start_text_region();
+        return;
+    }
+
+    if (m_mouse_is_over_text) {
+        m_is_dragging = true;
+        m_drag_start_point = event.layer_event().position();
+    } else {
+        // User clicked somewhere outside the currently edited text region
+        // apply the current text and then start a new one where they clicked.
+        apply_text_to_layer();
+        reset_tool();
+        start_text_region();
+    }
+}
+
+GUI::Widget* TextTool::get_properties_widget()
+{
+    if (!m_properties_widget) {
+        m_properties_widget = GUI::Widget::construct();
+        m_properties_widget->set_layout<GUI::VerticalBoxLayout>();
+
+        auto& font_header = m_properties_widget->add<GUI::Label>("Current Font:");
+        font_header.set_text_alignment(Gfx::TextAlignment::CenterLeft);
+
+        m_font_label = m_properties_widget->add<GUI::Label>(m_selected_font->human_readable_name());
+
+        auto& change_font_button = m_properties_widget->add<GUI::Button>("Change Font...");
+        change_font_button.on_click = [&](auto) {
+            auto picker = GUI::FontPicker::construct(nullptr, m_selected_font, false);
+            if (picker->exec() == GUI::Dialog::ExecResult::OK) {
+                m_font_label->set_text(picker->font()->human_readable_name());
+                m_selected_font = picker->font();
+                m_text_editor->set_font(m_selected_font);
+                m_editor->set_focus(true);
+            }
+        };
+    }
+
+    return m_properties_widget.ptr();
```

Comment:
Stylistic suggestion (feel free to ignore) - the diff looks odd, so move the if negated to the top, dedent the rest.

```suggestion
    if (m_properties_widget)
        return m_properties_widget.ptr();
    
    m_properties_widget = GUI::Widget::construct();
    m_properties_widget->set_layout<GUI::VerticalBoxLayout>();

    auto& font_header = m_properties_widget->add<GUI::Label>("Current Font:");
    font_header.set_text_alignment(Gfx::TextAlignment::CenterLeft);

    m_font_label = m_properties_widget->add<GUI::Label>(m_selected_font->human_readable_name());

    auto& change_font_button = m_properties_widget->add<GUI::Button>("Change Font...");
    change_font_button.on_click = [&](auto) {
        auto picker = GUI::FontPicker::construct(nullptr, m_selected_font, false);
        if (picker->exec() == GUI::Dialog::ExecResult::OK) {
            m_font_label->set_text(picker->font()->human_readable_name());
            m_selected_font = picker->font();
            m_text_editor->set_font(m_selected_font);
            m_editor->set_focus(true);
        }
    };
    return m_properties_widget.ptr();
```

Reply:
This is following the pattern that is in the other tools but I agree it looks better the other way around. I've made this change in the latest push